### PR TITLE
Json zombie streams

### PR DESCRIPTION
**Pull request by Wizarth**
Fixes https://github.com/strongloop/strong-remoting/issues/482
Includes test case.

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by bajtos** [comment]
Thank you @Wizarth for the pull request, I appreciate the test case! I'll try to review the changes in the next few (working) days.

**Comment by bajtos** [comment]
@slnode ok to test

File: test/streams.js [hunk 1]
```diff
@@ -180,6 +180,27 @@ describe('a function returning a ReadableStream', function() {
...
+    it('should close server stream on client disconnect',
+      function(done) {
+        const r = request(app)
+          .get('/StreamClass/createInfiniteStream')
+          .set('Accept', 'application/json')
+          // .set('Accept', 'text/event-stream')
+          .timeout(100)
+          .buffer(false);
+        r
+          .then(function() {
```
Comment (bajtos) [inline test/streams.js:193]: Mixing promises (`.then()` API) with callbacks (`done` argument) is usually a recipe for a disaster. I'll take a look on how to rewrite this test to be either fully callback-based or fully promise-based.

BTW when I run this new test agains the current master, the test fails with the following error:

```
 1) a function returning a ReadableStream
       an http client requesting a stream as JSON
         should close server stream on client disconnect:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/bajtos/src/strong/remoting/test/streams.js)
```

I am not sure if that's intended?

Comment (bajtos) [reply test/streams.js:193]: I managed to find a better way how to test this, will push the changes to your feature branch soon.

**Comment by bajtos** [comment]
@Wizarth Thank you again for the pull request. I added a new commit f7c066c where I have rewritten your test to use Node.js core API `http.request`, which gives us more control. Now the test returns a descriptive error message when the server event stream was not closed within 200ms (which is configurable).

I was not able to get rid of the mixed promise/callback style because of the way how `streamClosed` was created. We could rewrite this part from Promises to a regular EventEmitter, but I feel it's not worth the effort now.

It allowed me to remove your 50ms timer for calling `req.abort()`, now we abort the request just after we receive the response from the server. On my machine, the new test completes in ~40ms now.

On the other hand, the new test code is a bit longer than your original version. I think the benefits of faster execution and more robust implementation are worth it.

What do you think?

If you are fine with my changes, then we can proceed to squash the commits and land the pull request. Let me know!

**Comment by bajtos** [comment]
Also added 4986c52 to fix the new test on Node.js 8.x

**Comment by Wizarth** [comment]
Looks good to me.

I based the test on the existing test using streamClosed, which looks like it would have used the timeout as failure.

Using the http client response event is much better than a timer, I've never used that module directly.

**Comment by bajtos** [comment]
The fix is available in `strong-remoting@3.16.2`, thank you for the contribution! ❤️

**Pull request merged**